### DISCUSSION TITLE
Applying a function row-wise using multiple columns as input

### DISCUSSION THREAD

**Discussion by cebaa** [Kotlin/dataframe#421] [Q&A]
Say I have:

```
val dfj = dataFrameOf(
   "j" to (1..20).toList(),
   "k" to (1..20).toList(),
)
```

and a function:

```
fun f(j: Int, k: Int) { ... }
```

I'd like to add another column `l` to the dataframe by applying function `f` to values of `j` and `k` in each row.

How would I do that?

**Comment by koperagen** [comment]
Hi. 
In the notebooks create another cell after declaring dfj and write this:
```
dfj.add("newCol") { f(j, k) }
```
It works because in this environment dataframe can generate extension properties API after cell invocation, so you have type safe access to columns in dataframe and cell in rows.

If you work in IntelliJ project, you could use `add` and one of the functions for getting row value available in its lambda, like this:
```
dfj.add("newCol") {
    // type arguments are inferred from `f` parameters
    f(getValue("j"), getValue("k")) 
}
```
Possible alternatives:
```
f("j"(), "k"())
// or specify expected types as type arguments
f("j"<Int>(), "k"<Int>())
```
